lightningd/plugin.c

```diff
@@ -130,6 +154,9 @@ struct plugin *plugin_register(struct plugins *plugins, const char* path TAKES,
 		if (streq(path, p_temp->cmd)) {
 			if (taken(path))
 				tal_free(path);
+		 	/* If added as "important", upgrade to "important".  */
+			if (important)
+				p_temp->important = true;
```

Comment:
@ZmnSCPxj why ? I don't think we want to do this. *(sorry for the post-merge review)*

Reply:
A user might have a directory-full of plugins added by `add-plugin-dir`.  Then they decide later that one of those plugins is so vital that `lightningd` must die rather than move on if the plugin dies, so marks it as `important-plugin`.  It would be difficult otherwise to promote a plugin to important.

Reply:
Yeah but this loop is to check the path of the plugin we register against already registered one:
```C
/* Don't register an already registered plugin */
	list_for_each(&plugins->plugins, p_temp, list) {
		if (streq(path, p_temp->cmd)) {
			if (taken(path))
				tal_free(path);
		 	/* If added as "important", upgrade to "important".  */
			if (important)
				p_temp->important = true;
                            return NULL;
		}
	}
```
You should not make the cursor used to loop `important` if the plugin you want to register is important.

Reply:
It is correct.  Consider the case where `lightningd.conf`:

```
plugin-dir=/path/to/plugins/
important-plugin=/path/to/plugins/really-important-plugin.sh
```

`--plugin-dir` calls into `add_plugin_dir`, which iterates over all executable files in the dir, and calls `plugin_register` for each file.  This causes `plugin_register` to construct a `struct plugin` for each file in the dir, and if the dir contains `really-important-plugin.sh`, then a `struct plugin` will be added to the `plugins->plugins`, but (since it has not processed `important-plugin` yet!) marked as unimportant.

Then the *later* `important-plugin` will again call into `plugin_register`, which will enter the loop, and find the existing `/path/to/plugins/really-important-plugin.sh` `struct plugin`.  So it now does the upgrade of this already-existing plugin to `important`.

Otherwise we would require that all important plugins be listed before all `plugin-dir` indications, and since we have a default `plugin-dir`, we cannot store important plugins there.

In the case where the `important-plugin` is not already existing in the list of plugins, we initialize `p->important = important` in the later allocation code of the function.

Reply:
What i mean is the above code leads to *all currently registered plugins being marked as important* if an important plugin is started twice and happens to be at the end of the list. Will open a PR to discuss it, but maybe i'm wrong and i'll remark it by testing it ..

Reply:
The `if (streq(path, p_temp->cmd))` and `return NULL` both prevent this from running on all registered plugins.

Reply:
Ok, got it, sorry for insisting!